File: app/src/main/java/cc/rome753/surfacepaint/opengl/model/Sphere.java
package cc.rome753.surfacepaint.opengl.model;

import android.opengl.GLES20;

import cc.rome753.surfacepaint.opengl.data.VertexArray;
import cc.rome753.surfacepaint.opengl.programs.TextureShaderProgram;

import static cc.rome753.surfacepaint.opengl.Constants.BYTES_PER_FLOAT;

public class Sphere {
    private static final int POSITION_COMPONENT_COUNT = 3;
    private static final int TEXTURE_COORDINATES_COMPONENT_COUNT = 2;
    private static final int COMPONENT_COUNT = POSITION_COMPONENT_COUNT + TEXTURE_COORDINATES_COMPONENT_COUNT;
    private static final int STRIDE = (POSITION_COMPONENT_COUNT + TEXTURE_COORDINATES_COMPONENT_COUNT) * BYTES_PER_FLOAT;

    float R = 0.5f;

    int layers = 100;
    int sectors = 360;
    float alpha = (float) (Math.PI / layers);
    float beta = (float) (Math.PI * 2 / sectors);

    float sUnit = 1f / sectors;
    float tUnit = 1f / layers;

    private final VertexArray vertexArray;

    public Sphere() {

        float vertices[] = new float[(layers + 1) * (sectors + 1) * COMPONENT_COUNT];

        int index = -1;
        for (int i = 0; i < layers + 1; i++) {
            float r = (float) (R * Math.sin(i * alpha));
            float y = (float) (R * Math.cos(i * alpha));

            float t = i * tUnit;
            for (int j = 0; j < sectors + 1; j++) {
                float z = (float) (r * Math.sin(j * beta));
                float x = (float) (r * Math.cos(j * beta));

                float s = 1f - j * sUnit;

                vertices[++index] = x;
                vertices[++index] = y;
                vertices[++index] = z;
                vertices[++index] = s;
                vertices[++index] = t;
            }
        }

        float stripVertices[] = getStripArray(vertices, layers + 1, sectors + 1, COMPONENT_COUNT);
        vertexArray = new VertexArray(stripVertices);
    }

    public static float[] getStripArray(float[] array, int rows, int cols, int stride) {
        float stripArray[] = new float[(rows - 1) * (cols * 2) * stride];

        for (int i = 0; i < rows - 1; i++) {
            for (int j = 0; j < cols; j++) {
                int srcPos = (i * cols + j) * stride;
                int destPos = srcPos * 2;
                System.arraycopy(array, srcPos, stripArray, destPos, stride);

                destPos = destPos + stride;
                srcPos = ((i + 1) * cols + j) * stride;
                System.arraycopy(array, srcPos, stripArray, destPos, stride);
            }
        }
        return stripArray;
    }

    public void bindData(TextureShaderProgram textureProgram) {
        vertexArray.setVertexAttrPointer(
                0,
                textureProgram.getPositionAttributeLocation(),
                POSITION_COMPONENT_COUNT,
                STRIDE);

        vertexArray.setVertexAttrPointer(
                POSITION_COMPONENT_COUNT,
                textureProgram.getTextureCoordinatesAttributeLocation(),
                TEXTURE_COORDINATES_COMPONENT_COUNT,
                STRIDE);
    }

    public void draw() {
        for (int i = 0; i < layers; i++) {
            GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP,
                    i * (sectors + 1) * 2, ((sectors + 1) * 2));
        }
    }
}

File: opengles3video/src/main/java/com/example/android/Sphere.kt
package com.example.android

import android.opengl.GLES20
import android.opengl.GLES30
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer

class Sphere {
    var R = 0.5f
    var layers = 100
    var sectors = 360
    var alpha = (Math.PI / layers).toFloat()
    var beta = (Math.PI * 2 / sectors).toFloat()
    var sUnit = 1f / sectors
    var tUnit = 1f / layers

    var vertexBuffer: FloatBuffer? = null
    var vao: IntArray = IntArray(1)

    fun draw() {
        GLES30.glBindVertexArray(vao[0])
        for (i in 0 until layers) {
            GLES20.glDrawArrays(
                GLES20.GL_TRIANGLE_STRIP,
                i * (sectors + 1) * 2, (sectors + 1) * 2
            )
        }
    }

    companion object {
        private const val BYTES_PER_FLOAT = 4
        private const val POSITION_COMPONENT_COUNT = 3
        private const val TEXTURE_COORDINATES_COMPONENT_COUNT = 2
        private const val COMPONENT_COUNT =
            POSITION_COMPONENT_COUNT + TEXTURE_COORDINATES_COMPONENT_COUNT
        private const val STRIDE =
            (POSITION_COMPONENT_COUNT + TEXTURE_COORDINATES_COMPONENT_COUNT) * BYTES_PER_FLOAT

        fun getStripArray(array: FloatArray, rows: Int, cols: Int, stride: Int): FloatArray {
            val stripArray = FloatArray((rows - 1) * (cols * 2) * stride)
            for (i in 0 until rows - 1) {
                for (j in 0 until cols) {
                    var srcPos = (i * cols + j) * stride
                    var destPos = srcPos * 2
                    System.arraycopy(array, srcPos, stripArray, destPos, stride)
                    destPos = destPos + stride
                    srcPos = ((i + 1) * cols + j) * stride
                    System.arraycopy(array, srcPos, stripArray, destPos, stride)
                }
            }
            return stripArray
        }
    }

    init {
        var vertices = FloatArray((layers + 1) * (sectors + 1) * COMPONENT_COUNT)
        var index = -1
        for (i in 0 until layers + 1) {
            val r = (R * Math.sin((i * alpha).toDouble())).toFloat()
            val y = (R * Math.cos((i * alpha).toDouble())).toFloat()
            val t = i * tUnit
            for (j in 0 until sectors + 1) {
                val z = (r * Math.sin((j * beta).toDouble())).toFloat()
                val x = (r * Math.cos((j * beta).toDouble())).toFloat()
                val s = 1f - j * sUnit
                vertices[++index] = x
                vertices[++index] = y
                vertices[++index] = z
                vertices[++index] = s
                vertices[++index] = t
            }
        }

        vertices = getStripArray(vertices, layers + 1, sectors + 1, COMPONENT_COUNT)


        //分配内存空间,每个浮点型占4字节空间
        vertexBuffer = ByteBuffer.allocateDirect(vertices.size * 4)
            .order(ByteOrder.nativeOrder())
            .asFloatBuffer()
        //传入指定的坐标数据
        vertexBuffer!!.put(vertices)
        vertexBuffer!!.position(0)
        vao = IntArray(1)
        GLES30.glGenVertexArrays(1, vao, 0)
        GLES30.glBindVertexArray(vao[0])
        val vbo = IntArray(1)
        GLES20.glGenBuffers(1, vbo, 0)
        GLES20.glBindBuffer(GLES20.GL_ARRAY_BUFFER, vbo[0])
        GLES20.glBufferData(
            GLES20.GL_ARRAY_BUFFER,
            vertices.size * 4,
            vertexBuffer,
            GLES20.GL_STATIC_DRAW
        )
        GLES20.glVertexAttribPointer(0, 3, GLES20.GL_FLOAT, false, 5 * 4, 0)
        GLES20.glEnableVertexAttribArray(0)
        GLES20.glVertexAttribPointer(1, 2, GLES20.GL_FLOAT, false, 5 * 4, 3 * 4)
        GLES20.glEnableVertexAttribArray(1)

    }
}